### PR TITLE
policy: Add Port Range Support for Policies 3/3

### LINKED ISSUE #16622: Port ranges in network policies
The new `endPort` field for K8s' NetworkPolicies is in [alpha in v1.21](https://kubernetes.io/docs/concepts/services-networking/network-policies/#targeting-a-range-of-ports) and likely to become stable soon. It allows policies to target a range of ports instead of a single port.

Supporting this in Cilium will require changes in the datapath.

Tasks:
* [x] Add port range management mechanism
  * https://github.com/cilium/cilium/pull/20633
* [x] Add support for parsing/embedding port ranges from the API into the policy engine
* [x] Extend datapath for port range support
* [x] Implement policy engine -> datapath map generation logic

### PR DESCRIPTION
**Pull request by nathanjsweet**
See commits.

Fixes: #16622

Note: I'll address all missing documentation in a separate PR.

This builds on the following PRs:
- https://github.com/cilium/cilium/pull/23885
- https://github.com/cilium/proxy/pull/373
- https://github.com/cilium/cilium/pull/29717
- https://github.com/cilium/cilium/pull/32430
- https://github.com/cilium/cilium/pull/32675

```release-note
policy: Add support for port ranges in network policies.
```

### CODE REVIEW

**Comment by nathanjsweet** [comment]
The "ci-e2e-upgrade" is a known flake that is being investigated by the CI team.

File: cilium-dbg/cmd/helpers.go
Comment (doniacld) [inline cilium-dbg/cmd/helpers.go:346]: nit: A little typo in the commit message `Enovy`

File: pkg/envoy/xds_server.go
Comment (derailed) [inline pkg/envoy/xds_server.go:1620]: Could there be scenarios where protocol is unknown?

**Review by derailed — APPROVED** [review]
@nathanjsweet Nice work!

File: pkg/policy/l4.go
Comment (jrajahalme) [inline pkg/policy/l4.go]: Looks like production code is only calling this with a `nil` map. If do, maybe remove the argument here and define another test-only helper that can take the map argument?

File: pkg/policy/l4.go
Comment (jrajahalme) [inline pkg/policy/l4.go]: What if two different port ranges result in overlapping sets of masked ports? We'd need to "use-count" them and delete only if count reaches `0`?

File: pkg/policy/l4.go
Comment (jrajahalme) [inline pkg/policy/l4.go]: What if delete is called for a range that was not upserted in the first place? Now length is going to be wrong?

Maybe the caller should never delete if not upserted previously? If so, maybe log an error if any one of the masked ports is not found in the trie?

File: pkg/policy/l4.go
Comment (jrajahalme) [inline pkg/policy/l4.go]: could be more efficient to use a struct of the three integers as the map key here? Would avoid all the string conversion related allocations.

File: pkg/policy/l4.go
Comment (jrajahalme) [inline pkg/policy/l4.go:1470]: maybe comment why the LPM lookup is the correct thing to do here rather than the exact lookup?

File: pkg/policy/repository_deny_test.go
Comment (jrajahalme) [inline pkg/policy/repository_deny_test.go]: Does not have a functional difference, but would be nice to have `policyDeny. Equals(t, expectedDenyPolicy)` here to have visual balance with the `Diff` call after.

File: pkg/policy/repository_deny_test.go
Comment (jrajahalme) [inline pkg/policy/repository_deny_test.go]: flip the maps for `Equals` here as well...

**Review by jrajahalme — COMMENTED** [review]
Huge lift with all the testing code changes! Some questions regarding potentially overlapping masked ports in the trie.

**Review by nebril — APPROVED** [review]
k8s LGTM

File: pkg/envoy/xds_server.go
Comment (nathanjsweet) [reply pkg/envoy/xds_server.go:1620]: Not yet, but that will be coming. I'll make a note to convert this into a default case, which it should be.

File: pkg/policy/l4.go
Comment (nathanjsweet) [reply pkg/policy/l4.go]: Good call, though it should be the references themselves because the filters aren't necessarily the same.

File: pkg/policy/l4.go
Comment (nathanjsweet) [reply pkg/policy/l4.go]: The exact range delete by portProtoKey sovles this.

File: pkg/policy/l4.go
Comment (nathanjsweet) [reply pkg/policy/l4.go:1470]: I renamed "Lookup" to "LongestPrefixMatch" instead.

**Comment by nathanjsweet** [comment]
[This branch](https://github.com/cilium/cilium-cli/compare/main...pr/nathanjsweet/add-port-range-netpol-tests) of the cilium-cli has connectivity tests that use port ranges. I ran it locally, and all tests passed.

**Comment by nathanjsweet** [comment]
[This draft PR](https://github.com/cilium/cilium/pull/33238) is a copy of this PR with an additional commit to use [the cilium CLI that has port-range tests](https://github.com/cilium/cilium-cli/pull/2605) in the CI tests. It's passing on all tests, and the port-range tests are being run.

**Review by doniacld — APPROVED** [review]
lgtm

File: pkg/maps/policymap/policymap.go
Comment (joestringer) [inline pkg/maps/policymap/policymap.go]: nit: The code is accurate here, but I found it confusing to call this a `mask`. I don't see it being used for any masking operations in the code. Isn't this a map from a prefix length to the highest value of the corresponding suffix?

File: pkg/policy/l4.go
Comment (joestringer) [inline pkg/policy/l4.go:1130]: Why does production code link against `*testing.T`? Is it possible to move that into a testing wrapper?

File: pkg/policy/l4.go
Comment (joestringer) [inline pkg/policy/l4.go:1170]: nit: Is this at all related to the k8s feature "Named Ports"? We might want to clarify this somehow (ideally with a name that doesn't overlap, or alternatively with a comment that describes the relationship with the k8s feature, if any).

File: pkg/policy/k8s/network_policy.go
Comment (joestringer) [inline pkg/policy/k8s/network_policy.go:36]: Awesome! Is there an e2e test we can now enable?

File: pkg/policy/l4.go
Comment (joestringer) [inline pkg/policy/l4.go]: nit: I'm a bit confused what the "additional entries" are or what "only the vagaries of indexing by range" means.

Is the following a true statement?
> "rangeMapLen counts the number of unique L4Filters in the `rangePortMap`"

File: pkg/policy/l4.go
Comment (joestringer) [inline pkg/policy/l4.go]: nits: If this is only used for testing, we could probably keep it in the testing code.

Spelling:
```suggestion
// InitL4PolicyMap creates an new L4PolicyMap, with an initial
// set of values. The initMap argument does not support ranges.
func InitL4PolicyMap(initMap map[string]*L4Filter) L4PolicyMap {
```

Typically we'd also use the parlance `NewFoo()` for creating a new `Foo`, rather than `Init(...)` which sounds like it initializes an existing object.

File: pkg/policy/l4.go
Comment (joestringer) [inline pkg/policy/l4.go]: nit: Stutter, spelling
```suggestion
// Delete an L4Filter from the index by protocol/port-endPort
```

File: pkg/policy/l4.go
Comment (joestringer) [inline pkg/policy/l4.go]:
```suggestion
// LongestPrefixMatch looks up an L4Filter by protocol/port that contains the port and protocol
```

File: pkg/policy/l4.go
Comment (joestringer) [inline pkg/policy/l4.go:1226]: What happens if the user specifies two overlapping port ranges? It seems like this currently will just overwrite the initial upsert in a non-deterministic way.

**Review by joestringer — COMMENTED** [review]
Nice work! A few questions (also highlighted below)

1. Is there some e2e testing we can now enable with this PR? Presumably there's some tests in the upstream k8s suite we've been skipping or something?
2. What happens if a user creates two overlapping policies with endports? Or combine a KNP port range with a CNP with L7? Is there some weird corner case condition that could have unexpected behaviour? Might be worth identifying such conditions if any and document them. (Thread in the last comment below)
3. Related, is this enabling only port ranges in KNP or also CNP (it's fine and probably ideal to limit to KNP for now, since CNP could introduce additional complications given the possibility of L7 fields in the L4 stanza)
4. It would be nice to move some of the testing-only code from l4.go out so we don't pull the testing package in as a runtime dependency.

EDIT 2024-06-20: (2) is resolved though could potentially be better tested, (4) is deferred to a later PR.

File: pkg/policy/l4.go
Comment (joestringer) [reply pkg/policy/l4.go:1226]: ^ If this is only used with KNP, then the L4Filter is probably equivalent if two overlapping port ranges are injected. Then I guess if a CNP with L7 is added for a specific port, then that will be held in the other map, not in this port ranges map. So long as we properly encode the map entry for the specific port, that aspect should work. So perhaps the only awkward case would be if we enabled port ranges in CNP and allowed two overlapping port ranges with L7 fields under the L4 stanza...?

Comment (jrajahalme) [reply pkg/policy/l4.go:1226]: It should not matter which range (or a part thereof) caused the initial `l4M.rangePortMap.Upsert` to be made, as the inner map will have a different key for each different range.

File: pkg/policy/l4.go
Comment (jrajahalme) [inline pkg/policy/l4.go]: It should be functionally equivalent to just `return` here. If not, then the l4PolicyMap was in an inconsistent state?

File: pkg/policy/l4.go
Comment (jrajahalme) [inline pkg/policy/l4.go:1285]: For a l4PolicyMap in a consistent state the same l4 should be found for each of the MaskedPorts in the range. So, to check that the map is in a consistent state we could check for exactly that. Or, if we trust the map to be in a consistent state (`return l4` for the first one found), we could simply `return nil` instead of `continue` for the same reason?

File: pkg/policy/l4.go
Comment (jrajahalme) [inline pkg/policy/l4.go:1313]: `shortestPortRange` should be updated here.

**Review by jrajahalme — CHANGES_REQUESTED** [review]
Small error of omission in `LongestPrefixMatch` that needs to be fixed.

Also some inconsistency on trusting the map to be in a consistent state (`continue` statements could be replaced with `return` statements, or then we could/should check that all masked ports actually exist in the outer map?

Could perhaps add testing code to fully validate the consistency of the outer map and take the shortcuts in production code?

File: pkg/policy/l4.go
Comment (nathanjsweet) [reply pkg/policy/l4.go:1130]: > Why does production code link against *testing.T? Is it possible to move that into a testing wrapper?

Because those methods are used by other packages. This PR does not introduce this dependency. It was introduced when ~`L4PolicyMap`~ `MapState` became an interface. The intent of introducing the `testing` package was to enforce that it only be used in tests, but we could rename the methods something like `EqualsTestingOnly` and `DiffTestingOnly`.

**Edit**: Referred to the wrong struct.

File: pkg/policy/l4.go
Comment (nathanjsweet) [reply pkg/policy/l4.go]: Yes. "The vagaries of indexing by range" is an opaque reference to the idea that we can't have a single entry for one range because bitwise math does not allow it, so we need to split the range up into range lengths (divisors) of powers of two. I will try to document this better.

File: pkg/policy/l4.go
Comment (nathanjsweet) [reply pkg/policy/l4.go]: >nits: If this is only used for testing, we could probably keep it in the testing code.
It's used by the envoy package, unfortunately.

> Typically we'd also use the parlance NewFoo() for creating a new Foo, rather than Init(...) which sounds like it initializes an existing object.

What about `InitNew` or `NewL4PolicyMapInit`?

File: pkg/policy/l4.go
Comment (nathanjsweet) [reply pkg/policy/l4.go]: True.

File: pkg/policy/l4.go
Comment (joestringer) [reply pkg/policy/l4.go:1130]: > This PR does not introduce this dependency. It was introduced when L4PolicyMap became an interface.

I'm confused. This comment thread is responding to the change for `L4PolicyMap` to convert it to an interface and add the `testing.T` dependency, as well as the implementations for this new interface.

I understand it might be a little less convenient to move this testing-only functionality into a testing-only file, but that seems better than entangling the direct testing dependency into the prod code like this. There's probably a reasonable wrapper we could create, for instance if we moved `InitL4PolicyMap()` into the testing code and forced it to return an object that (a) embeds L4PolicyMap and (b) implements the additional testing-only functions.

EDIT: I guess this particular proposal doesn't fit with the suggestion for using builder pattern below, but there's probably a reasonable solution either by avoiding builder pattern or by creating a testing-only wrapper around the implementation with builder pattern.

File: pkg/policy/l4.go
Comment (joestringer) [reply pkg/policy/l4.go]: I guess the common pattern would be something like the following:

`func (l4 L4PolicyMap) WithFilters(initMap map[string]*L4Filter) L4PolicyMap { ... }` 

Then we could use builder pattern elsewhere, eg:

`pm := NewL4PolicyMap().WithFilters(...)`

File: pkg/policy/l4.go
Comment (joestringer) [reply pkg/policy/l4.go:1226]: What if the range is the same, but the L7 parameters are different? Will they be merged as you might expect?

File: pkg/policy/l4.go
Comment (nathanjsweet) [reply pkg/policy/l4.go:1130]: > I'm confused. This comment thread is responding to the change for L4PolicyMap to convert it to an interface and add the testing.T dependency, as well as the implementations for this new interface.

Ah sorry, my bad. It wasn't L4PolicyMap, but `MapState` that introduced a testing dependency into the package with its `Diff` method. 

> I guess this particular proposal doesn't fit with the suggestion for using builder pattern below, but there's probably a reasonable solution either by avoiding builder pattern or by creating a testing-only wrapper around the implementation with builder pattern.

Since the cat is already out of the bag I would prefer to remove the `testing` package dependency out in a different PR.

File: pkg/policy/l4.go
Comment (nathanjsweet) [reply pkg/policy/l4.go:1226]: >What if the range is the same, but the L7 parameters are different? Will they be merged as you might expect?

Yes. In the `createFilter` method an exact range match is required to merge filters. We always know that unique ranges and ports are being upserted.

File: pkg/policy/l4.go
Comment (nathanjsweet) [reply pkg/policy/l4.go:1170]: Updated with comment.

File: pkg/maps/policymap/policymap.go
Comment (nathanjsweet) [reply pkg/maps/policymap/policymap.go]: Renamed and updated comment to be more clear.

File: pkg/policy/l4.go
Comment (joestringer) [reply pkg/policy/l4.go:1130]: Ah right, that makes more sense. Fair enough, we can follow up on this.

File: pkg/policy/l4.go
Comment (joestringer) [reply pkg/policy/l4.go:1226]: Ah right yes, handled in the layers above. :+1: 

```
INCOMING CALLS
- f Upsert github.com/cilium/cilium/pkg/policy • l4.go
  - f addL4Filter github.com/cilium/cilium/pkg/policy • l4.go
    - f mergeIngressPortProto github.com/cilium/cilium/pkg/policy • rule.go
      - f mergeIngress github.com/cilium/cilium/pkg/policy • rule.go
        + f resolveIngressPolicy github.com/cilium/cilium/pkg/policy • rule.go
```

Comment (joestringer) [reply pkg/policy/l4.go:1226]: I also looked into another corner case, if the start port is the same but the endport is different. That appears to be covered by indexing this map with a key of {port,endPort,proto}. I also see that `TestL4PolicyMapPortRangeOverlaps()` validates the map management aspects here with overlapping ranges. :+1: 

I didn't see unit tests for the outer layer that includes merging ranges with L7 fields. Maybe that's an area for subsequent improvement to reduce the risk of regression specifically related to those features. There are probably other testing aspects we should focus on first though.

File: pkg/policy/l4.go
Comment (joestringer) [inline pkg/policy/l4.go]: Should this also check `.Ingress`?

File: pkg/policy/l4.go
Comment (joestringer) [inline pkg/policy/l4.go:1130]: nit: We should probably look at which of these actually need to be public, and hide them by default. Upsert directly here rather than via `addL4Filter()` would bypass some rule merging logic, Delete could delete filters that are actively used by another rule, etc.

File: pkg/policy/l4.go
Comment (joestringer) [inline pkg/policy/l4.go:1185]: nit: One of these days, we should probably trace these all the way up the stack and just sanitize the types at creation to avoid type conversions all the way deep in policy generation like this ;-)

**Review by joestringer — APPROVED** [review]
Minor nits below, but LGTM. I'd still like to better understand the plan for testing, comment thread hidden above.

File: pkg/policy/l4.go
Comment (nathanjsweet) [reply pkg/policy/l4.go:1130]: [Jarno's PR](https://github.com/cilium/cilium/pull/33302) addresses this.

File: pkg/policy/k8s/network_policy.go
Comment (nathanjsweet) [reply pkg/policy/k8s/network_policy.go:36]: I couldn't find anything. If you come across something, please let me know. The [Kubernetes test](https://github.com/cilium/cilium/blob/main/test/kubernetes-test.sh) does not have exclusions for port ranges.

**Review by jrajahalme — APPROVED** [review]

File: pkg/policy/k8s/network_policy.go
Comment (joestringer) [reply pkg/policy/k8s/network_policy.go:36]: :raised_eyebrow: There must be tests somewhere, it's [stable in k8s since late 2022](https://kubernetes.io/blog/2022/08/23/kubernetes-v1-25-release/#promoted-endport-in-network-policy-to-stable). We shouldn't be merging functionality like this without a plan for automated testing.

File: pkg/policy/l4.go
Comment (joestringer) [reply pkg/policy/l4.go:1130]: I don't understand how that PR is related. The linked PR seems to be providing stronger guarantees at a subsequent policy stage related to EndpointPolicy and map state.

`addL4Filter()` does a merge L4filter + call into the publicly exported `Upsert` from this API. If ever someone in future decided to attempt to directly `Upsert` using this API in this interface, then they would be avoiding the filter merge logic. This is not a huge deal because if someone attempted to propose such a patch, we would also review that code and hopefully catch the discrepancy. I just point it out as a bit of a code smell that we shouldn't be exporting functions that are not intended for use outside the package.

**Comment by nathanjsweet** [comment]
https://github.com/cilium/cilium/pull/33238 is updated and green as well. This PR is ready to merge until it approved for release in 1.16 or the 1.16 release happens.

**Comment by joestringer** [comment]
Summarazing the [Slack discussion](https://cilium.slack.com/archives/C2B917YHE/p1718992983793069) here: This feature was substantially completed already prior to feature freeze going into effect, and had just relatively minor feedback remaining prior to the date. The feature has been under high demand per the linked issue for at least a couple of years and we've finally got an implementation that the Cilium committers are happy with here. We had multiple committers and involved community members from different organizations advocating for a feature freeze exception for this feature, so it is included in the v1.16 release. We haven't yet branched off for v1.16 maintenance cycle, so the option is open for inclusion without additional backporting. Users are already taking mitigative actions due to the lack of support in Cilium. Assuming that we continue to validate & test the implementation through until release, it seems reasonable to grant an exception here and merge the PR during the feature freeze.

**Pull request merged**
